cabot_navigation2/plugins/cabot_dwb_local_planner.cpp

```diff
@@ -0,0 +1,208 @@
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+nav_2d_msgs::msg::Path2D
+CaBotDWBLocalPlanner::transformGlobalPlan(
+  const nav_2d_msgs::msg::Pose2DStamped & pose)
+{
+  if (global_plan_.poses.empty()) {
+    throw nav2_core::PlannerException("Received plan with zero length");
+  }
+
+  // let's get the pose of the robot in the frame of the plan
+  nav_2d_msgs::msg::Pose2DStamped robot_pose;
+  if (!nav_2d_utils::transformPose(
+      tf_, global_plan_.header.frame_id, pose,
+      robot_pose, transform_tolerance_))
+  {
+    throw dwb_core::
+          PlannerTFException("Unable to transform robot pose into global plan's frame");
+  }
+
+  // we'll discard points on the plan that are outside the local costmap
+  nav2_costmap_2d::Costmap2D * costmap = costmap_ros_->getCostmap();
+  double dist_threshold = std::max(costmap->getSizeInCellsX(), costmap->getSizeInCellsY()) *
+    costmap->getResolution() / 2.0;
+
+  // If prune_plan is enabled (it is by default) then we want to restrict the
+  // plan to distances within that range as well.
+  double prune_dist = prune_distance_;
+
+  // Set the maximum distance we'll include points before getting to the part
+  // of the path where the robot is located (the start of the plan). Basically,
+  // these are the points the robot has already passed.
+  double transform_start_threshold;
+  if (prune_plan_) {
+    transform_start_threshold = std::min(dist_threshold, prune_dist);
+  } else {
+    transform_start_threshold = dist_threshold;
+  }
+
+  // Set the maximum distance we'll include points after the part of the plan
+  // near the robot (the end of the plan). This determines the amount of the
+  // plan passed on to the critics
+  double transform_end_threshold;
+  double forward_prune_dist = forward_prune_distance_;
+  if (shorten_transformed_plan_) {
+    transform_end_threshold = std::min(dist_threshold, forward_prune_dist);
+  } else {
+    transform_end_threshold = dist_threshold;
+  }
+
+  // Customized: Find the closest point from the robot pose
+  auto nearest_point_from_robot = std::min_element(
+    global_plan_.poses.begin(), global_plan_.poses.end(),
+    [&](const auto & pose1, const auto & pose2) {
+      return euclidean_distance(pose1, robot_pose.pose) < euclidean_distance(pose2, robot_pose.pose);
+    });
+  double nearest_distance = euclidean_distance(*nearest_point_from_robot, robot_pose.pose);
+
+  // Customized: Find the first pose less than double of nearset_distance - adhoc solution
+  auto search_start_point = std::find_if(
+    global_plan_.poses.begin(), global_plan_.poses.end(),
+    [&](const auto & pose) {
+      return euclidean_distance(pose, robot_pose.pose) < nearest_distance * 2;
+    });
```

Comment:
What is the purpose of adjusting search_start_point from nearest_point_from_robot?

Reply:
In the original logic, the path's poses are observed from the beginning, and the process stops when the pose is farther than the threshold distance. There was an issue where if the initial pose is within the threshold circle, but the path subsequently moves outside the circle, the target point was set to the first pose outside the circle. (If the red line in your figure starting within the circle.)

<img width="713" alt="Screenshot 2024-07-03 at 11 54 25 PM" src="https://github.com/CMU-cabot/cabot-navigation/assets/43101027/339c6562-aa31-41ca-bc4a-8fab009a3016">

While drawing this, I found that the current logic may not work in some cases...